crates/matrix-sdk-crypto/src/olm/group_sessions/sender_data_finder.rs

```diff
@@ -0,0 +1,891 @@
+        // And is the signature in the device info valid?
+        let maybe_msk_info =
+            self.msk_if_device_is_signed_by_user(&sender_device, &sender_user_identity).await;
+
+        if let Some((msk, msk_verified)) = maybe_msk_info {
+            // Yes: H (cross-signing key matches that used to sign the device info!)
+            // and: J (device info is verified by matching cross-signing key)
+
+            // Find the actual key within the MasterPubkey struct
+            if let Some(msk) = msk.get_first_key() {
+                // We have MXID and MSK for the user sending the to-device message.
+                // Decide the MSK trust level based on whether we have verified this user.
+                // Set the MXID, MSK and trust level in the session. Remove the device
+                // info and retries since we don't need them.
+                Ok(SenderData::SenderKnown {
+                    user_id: sender_device.user_id().to_owned(),
+                    msk,
+                    msk_verified,
+                })
+                // TODO: [drop the lock]
+            } else {
+                // Surprisingly, there was no key in the MasterPubkey. We did not expect this:
+                // treat it as if the device was not signed by this master key.
+                //
+                tracing::error!(
+                    "MasterPubkey for user {} does not contain any keys!",
+                    msk.user_id()
+                );
+
+                Ok(SenderData::DeviceInfo {
+                    device_keys: sender_device.as_device_keys().clone(),
+                    retry_details: SenderDataRetryDetails::retry_soon(),
+                    legacy_session: true, // TODO: change to false when retries etc. are done
+                })
+                // TODO: [drop the lock]
+            }
+        } else {
+            // No: Device was not signed by the known identity of the sender.
+            // (Or the signature was invalid. We don't know which unfortunately, so we can't
+            // bail out completely if the signature was invalid, as we'd like
+            // to.)
+            //
+            // Since we've already checked there are >1 signatures, we guess
+            // it was signed by a different identity of this user, so we should retry
+            // later, in case either the device info or the user's identity changes.
+            Ok(SenderData::DeviceInfo {
+                device_keys: sender_device.as_device_keys().clone(),
+                retry_details: SenderDataRetryDetails::retry_soon(),
+                legacy_session: true, // TODO: change to false when retries etc. are done
+            })
+
+            // TODO: Increment retry_count and set
+            // next_retry_time_ms per backoff algorithm; let the background job
+            // pick it up [drop the lock]
+        }
+    }
+
+    /// If sender_device is correctly cross-signed by user_identity, return
```

Comment:
Are you referencing the arguments here? If so, let's highlight that and use the correct name.
```suggestion
    /// If `sender_device` is correctly cross-signed by `sender_user_identity`, return
```